### PR TITLE
bpo-31109: Convert zipimport to argument clinic

### PR DESCRIPTION
**Pull request by jarondl**
Convert the zipimport module to argument clinic. Mostly
straight-forward, had to rename a few argument names.

https://bugs.python.org/issue31109

### CODE REVIEW

File: Modules/zipimport.c [hunk 1]
```diff
@@ -54,46 +54,72 @@ static PyObject *get_data(PyObject *archive, PyObject *toc_entry);
...
 
 /* zipimporter.__init__
    Split the "subdirectory" from the Zip archive path, lookup a matching
    entry in sys.path_importer_cache, fetch the file directory from there
    if found, or else read it from the archive. */
+
+/*[clinic input]
+zipimport.zipimporter.__init__
+
+    archivepath: object(converter="PyUnicode_FSDecoder")
```
Comment (serhiy-storchaka) [inline Modules/zipimport.c]: Why not use just the name "path"?

File: Modules/zipimport.c [hunk 2]
```diff
@@ -379,21 +405,29 @@ find_loader(ZipImporter *self, PyObject *fullname, PyObject **namespace_portion)
     return FL_MODULE_FOUND;
 }
 
+/*[clinic input]
+zipimport.zipimporter.find_module
+
+    fullname: unicode
```
Comment (serhiy-storchaka) [inline Modules/zipimport.c:411]: The arguments of this method (and many other methods) were positional-only. Converting to Argument Clinic shouldn't change the behavior.

Comment (jarondl) [reply Modules/zipimport.c:411]: Actually *all* the methods here were positional-only. But this change is backwards compatible and logical. See for example http://bugs.python.org/issue20294#msg208444
Why should we force positional-only? Your patch in bpo-25711 to convert `zipimport` to Python would also have the same effect.

File: Modules/zipimport.c (hunk 1 above)
Comment (jarondl) [reply Modules/zipimport.c]: Do you mean `path:` or `archivepath as path:`? 
Sure I can do the latter, but what's the benefit? It was more straight forward to rename the variable to match the signature. The meaning of `archivepath` is clearer

File: Modules/zipimport.c (hunk 2 above)
Comment (jarondl) [reply Modules/zipimport.c:411]: Do you prefer to keep Argument Clinic conversion separate from behavioral changes? Then we can add trailing `/`s to the argument lists and remove them in a separate PR.

Comment (brettcannon) [reply Modules/zipimport.c:411]: Yeah, we prefer to not change semantics while doing an Argument Clinic conversion (we have run into issues in the past when doing both at the same time).

Comment (jarondl) [reply Modules/zipimport.c:411]: OK, done, now they are positional only. Thank you both for reviewing.

File: Modules/zipimport.c (hunk 1 above)
Comment (brettcannon) [reply Modules/zipimport.c]: The benefit of not renaming is the patch is much easier to review that way. As it stands it's much harder for us to check the accuracy of the changes due to so many lines being flagged as changed compared to just reviewing the Argument Clinic changes which is what this PR is specifically targeting.

**Review by brettcannon — CHANGES_REQUESTED** [review]

File: Modules/zipimport.c (hunk 1 above)
Comment (jarondl) [reply Modules/zipimport.c]: Done.

File: Modules/zipimport.c [hunk 3]
```diff
@@ -54,28 +54,54 @@ static PyObject *get_data(PyObject *archive, PyObject *toc_entry);
...
+    /
+
+Create a new zipimporter instance.
+
+'archivepath' must be a path-like object to a zipfile, or to a specific path
+inside a zipfile. For example, it can be '/tmp/myimport.zip', or
+'/tmp/myimport.zip/mydirectory', if mydirectory is a valid directory inside
+the archive.
+
+'ZipImportError is raised if 'archivepath' doesn't point to a valid Zip
```
Comment (brettcannon) [inline Modules/zipimport.c]: Looks like there is a stray `'` at the start of this sentence.

File: Modules/zipimport.c [hunk 4]
```diff
@@ -54,28 +54,54 @@ static PyObject *get_data(PyObject *archive, PyObject *toc_entry);
...
+archive.
+
+The 'archive' attribute of zipimporter objects contains the name of the
```
Comment (brettcannon) [inline Modules/zipimport.c]: "zipimporter objects" -> "the zipimporter object".

File: Modules/zipimport.c [hunk 5]
```diff
@@ -416,23 +451,32 @@ zipimporter_find_module(PyObject *obj, PyObject *args)
...
+/*[clinic input]
+zipimport.zipimporter.find_loader
+
+    fullname: unicode
+    path: object = None
+    /
+
+Search for a module specified by 'fullname'.
+
+'fullname' must be the
```
Comment (brettcannon) [inline Modules/zipimport.c]: Premature line break.

**Review by brettcannon — CHANGES_REQUESTED** [review]
Three really minor formatting tweaks necessary and then I think this will be ready to go!

**Review by brettcannon — APPROVED** [review]
@serhiy-storchaka did you have any other feedback or can I merge this?

**Comment by jarondl** [comment]
@serhiy-storchaka and @brettcannon , is there anything else I can do to have this merged?
In any case I want to say thank you both for the reviews.

**Comment by brettcannon** [comment]
Since it looks like @serhiy-storchaka has nothing to add would you mind adding a news entry? See https://devguide.python.org/committing/#what-s-new-and-news-entries on how to do that. Once that's file in I can touch it up as necessary and then merge this!

**Review by brettcannon — CHANGES_REQUESTED** [review]
Please add a news entry.

**Comment by jarondl** [comment]
Blimey, I didn't expect the Spanish Inquisition!

**Review by brettcannon — APPROVED** [review]
Once Travis goes green I'll merge!

**Comment by brettcannon** [comment]
Thanks for the work, @jarondl !

**Pull request merged**